### ISSUE TITLE
Roadie is erratically causing emails to be encoded differently

### ISSUE THREAD

**Issue by peternixey** [Mange/roadie#132]
There is a chance that this is not a Roadie issue in which case I apologise in advance and can take it elsewhere. But I do know that the presence of certain styles causes the issue to flip on and off.

The problem
---------
Certain styles encoded by Roadie are causing mail encoding to flip. In both instances below, Roadie is parsing the email, however in one case all styles are commented out and in the other the styles are active. 

**Sample of an email with all styles commented out**

````
Content-Type: text/html;\r\n
 charset=UTF-8\r\n
Content-Transfer-Encoding: 7bit\r\n
\r\n
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional //EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\r\n
<html>\r\n
<head>\r\n
<!-- attempt to stop apple auto-linking dates in emails -->\r\n
<meta content="date=no" name="format-detection">\r\n
<meta content="width=device-width" maximum-scale="1" minimum-scale="1" name="viewport" user-scalable="no">\r\n
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">\r\n
</head>\r\n
<body>\r\n
\r\n
<div class="list-post-layout">\r\n
<div class="preview-text-component">\r\n
Et ut velit totam enim dolore eius quod cupiditate. Enim illo similique recusandae nulla facere quo non. Consequatur perspiciatis omnis voluptas voluptatem.\r\n
</div>\r\n
\r\n
<div class="author-header-component">\r\n
<table>\r\n
<tr>\r\n
<td class="picture">\r\n
<span class="medium profile-picture-component">\r\n
<a href="http://example.com/incidunt-eius-dolor/members/person4?uid=28855108"><img alt="Pic" width="30" height="30" src="https://s3-eu-west-1.amazonaws.com/copyintest/profile_imgs/missing/medium.jpeg">\r\n
</a></span>\r\n
\r\n
</td>
````

*note - I've added actual linebreaks to make it easier to read but kept all line-break encodings (as printed into the console) to highlight what's happening**

**Sample of an email that includes styles**

````
Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: quoted-printable\r\n\r\n
<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional //EN\" \"http://w=\r\nww.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\r\n<html>\r\n<head>\r\n<!-- attempt to stop apple auto-linking dates in emails -->\r\n<meta content=3D\"date=3Dno\" name=3D\"format-detection\">\r\n<meta content=3D\"width=3Ddevice-width\" maximum-scale=3D\"1\" minimum-scale=3D=\r\n\"1\" name=3D\"viewport\" user-scalable=3D\"no\">\r\n<meta http-equiv=3D\"Content-Type\" content=3D\"text/html; charset=3DUTF-8\">=\r\n\r\n
<style>.reply-privately-button-component a:hover{color:#48B252}</style>\r\n</head>\r\n<body style=3D\"font-family:Helvetica, arial\">\r\n\r\n
<div class=3D\"list-post-layout\" style=3D\"padding-top:10px\">\r\n<div class=3D\"preview-text-component\" style=3D\"position:absolute;font-siz=\r\ne:1px;mso-hide:all;visibility:hidden;opacity:0;overflow:hidden;top:-10px;=\r\nleft:-10px;height:1px;width:1px;color:white;line-height:1px;max-height:0p=\r\nx;max-width:0px;display:none !important\">\r\nIn ut sint corporis voluptatem rerum cumque fuga quo. Qui ut maxime maior=\r\nes possimus omnis laudantium. Nam ea voluptatem omnis culpa veritatis tem=\r\npora commodi qui.\r\n</div>\r\n\r\n
<div class=3D\"author-header-component\" style=3D\"margin-bottom:5px;font-si=\r\nze:0.9rem\">\r\n<table style=3D\"border-collapse:collapse;width:100%\">\r\n<tr>\r\n<td class=3D\"picture\" style=3D\"border-collapse:collapse;width:40px\">\r\n<span class=3D\"medium profile-picture-component\" style=3D\"display:inline-=\r\nblock;position:relative;font-size:20px\">\r\n<a href=3D\"http://example.com/ipsam-est-dolorum/members/person12?uid=3D32=\r\n690103\" style=3D\"color:#48B252;text-decoration:none\"><img alt=3D\"Pic\" wid=\r\nth=3D\"30\" height=3D\"30\" src=3D\"https://s3-eu-west-1.amazonaws.com/copyint=\r\nest/profile_imgs/missing/medium.jpeg\" style=3D\"border:0;outline:none;text=\r\n-decoration:none;border-radius:50%\">\r\n</a></span>\r\n\r\n
</td>\r\n
````

Note the presence of the `=3D` (instead of just an equals) and the `=\r\n` character sequences. 

When running this in tests the switch to the encoding is actually erratic. It seems like it *might* be triggered by length of the email as my test suite generates random words for it. 

I'd be really grateful if you can give any thoughts on why this is happening or how to prevent it. Thank you!

**Comment by Mange** [comment]
It's likely that this is an effect from Nokigiri parsing the document and
then generating the new markup from the DOM.

Try to pipe your untouched markup to Nokigiri's HTML parser and then
inspect the output from #to_html on the DOM.

Is it happening there as well? If so, it's not directly caused by Roadie.
If not, I suspect it has something to do with the contents of your CSS, but
what that could be I don't know yet.

I noticed that the Content-Transfer-Encoding header changed, so the 3D part
probably comes from that.
What are you doing with the markup here? Are you passing it to Mail, or
sending it over HTTP to a browser? Are you using a framework in that case?

Den tis 20 dec. 2016 14:36Peter Nixey <notifications@github.com> skrev:

**Comment by peternixey** [comment]
Hi Mange and thanks for replying so quickly. I was thinking on the way into work that there was one aspect that made me think it wasn't Nokogiri and that is that the line-breaks that are being added are at a consistent length. When you swap out all the encoded line breaks  for actual line breaks you find that each line is exactly the same length. 

Your comment about the encoding got me thinking though and I noticed that at the top of each of the snippets was an encoding instruction. One was `Content-Transfer-Encoding: 7bit` and the other is `Content-Transfer-Encoding: quoted-printable`. 

I looked the two of these up and apparently quoted-printable exists because there is a [1000 character line-length limit](https://noobsnippets.wordpress.com/2015/06/20/quoted-printable-crossing-the-1000-character-smtp-barrier/) in SMTP. [Quoted printable](https://en.wikipedia.org/wiki/Quoted-printable) adds soft line breaks as well as encoding `=` characters as `=3D`. 

So what was happening was that as the syles were added in by Roadie, certain lines pushed out over 1000 characters long and Rails (which to answer your question is the framework I'm using) apparently flipped from 7bit to quoted-printable. 

For testing purposes, the way to then examine the email without the encoding is to access it via `mail.html_part.decoded.to_s` rather than `mail.html_part.to_s` which is what I was previously doing. 

Thanks again for the help and hope this also helps someone else too.
